File: python/questions/isBalanced.py
from typing import Optional

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        def height(node: Optional[TreeNode]) -> int:
            if node is None:
                return 0
            l = height(node.left)
            r = height(node.right)
            if l==-1 or r==-1 or abs(l-r)>1:
                return -1
            return max(l, r) + 1
        return height(root) != -1

File: cpp/questions/isBalanced.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    bool isBalanced(TreeNode* root) {
        auto height = [](this auto&& height, TreeNode* node) -> int {
            if (node == nullptr) {
                return 0;
            }
            int l = height(node->left);
            int r = height(node->right);
            if (l==-1 || r==-1 || std::abs(l-r)>1) {
                return -1;
            }
            return std::max(l, r) + 1;
        };
        return height(root) != -1;
    }
};
